File: CS/WpfApplication15/ViewModel.cs
using System;
using System.Linq;
using System.ComponentModel;
using System.Data;
namespace WpfApplication15 {
    public class ViewModel : INotifyPropertyChanged {
        DataSet DataSet;
        object _masterSource;
        object _detailSource;
        DataRowView _currentRow;

        public object MasterSource {
            get { return _masterSource; }
            private set {
                if (_masterSource == value) return;
                _masterSource = value;
                RaisePropertyChanged("MasterSource");
            }
        }
        public object DetailSource {
            get { return _detailSource; }
            private set {
                if (_detailSource == value) return;
                _detailSource = value;
                RaisePropertyChanged("DetailSource");
            }
        }
        public DataRowView CurrentRow {
            get { return _currentRow; }
            set {
                if (_currentRow == value) return;
                _currentRow = value;
                RaisePropertyChanged("CurrentRow");
                UpdateDetailSource();
            }
        }

        public ViewModel() {
            DataSet = CreateData();
            MasterSource = DataSet.Tables["Company"];
        }

        DataSet CreateData() {
            DataTable mdt = new DataTable("Company");
            mdt.Columns.Add(new DataColumn("Name", typeof(string)));
            mdt.Columns.Add(new DataColumn("ID", typeof(int)));
            mdt.Rows.Add("Ford", 4);
            mdt.Rows.Add("Nissan", 5);
            mdt.Rows.Add("Mazda", 6);

            DataTable ddt = new DataTable("Models");
            ddt.Columns.Add(new DataColumn("Name", typeof(string)));
            ddt.Columns.Add(new DataColumn("MaxSpeed", typeof(int)));
            ddt.Columns.Add(new DataColumn("CompanyName", typeof(string)));
            ddt.Rows.Add("FordFocus", 400, "Ford");
            ddt.Rows.Add("FordST", 400, "Ford");
            ddt.Rows.Add("Note", 1000, "Nissan");
            ddt.Rows.Add("Mazda3", 1000, "Mazda");

            DataSet ds = new DataSet("CM");
            ds.Tables.Add(mdt);
            ds.Tables.Add(ddt);
            DataRelation dr = new DataRelation("CompanyModels", mdt.Columns["Name"], ddt.Columns["CompanyName"]);
            ds.Relations.Add(dr);

            return ds;
        }
        void UpdateDetailSource() {
            DataView dv = new DataView(DataSet.Tables["Models"]);
            dv.RowFilter = String.Format("CompanyName = '{0}'", CurrentRow["Name"]);
            DetailSource = DataSet.Tables["Models"];
        }

        public event PropertyChangedEventHandler PropertyChanged;
        void RaisePropertyChanged(string propName) {
            if (PropertyChanged != null) PropertyChanged(this, new PropertyChangedEventArgs(propName));
        }
    }
}

File: VB/WpfApplication15/ViewModel.vb
Imports System
Imports System.Linq
Imports System.ComponentModel
Imports System.Data
Namespace WpfApplication15
    Public Class ViewModel
        Implements INotifyPropertyChanged

        Private DataSet As DataSet
        Private _masterSource As Object
        Private _detailSource As Object
        Private _currentRow As DataRowView

        Public Property MasterSource() As Object
            Get
                Return _masterSource
            End Get
            Private Set(ByVal value As Object)
                If _masterSource Is value Then
                    Return
                End If
                _masterSource = value
                RaisePropertyChanged("MasterSource")
            End Set
        End Property
        Public Property DetailSource() As Object
            Get
                Return _detailSource
            End Get
            Private Set(ByVal value As Object)
                If _detailSource Is value Then
                    Return
                End If
                _detailSource = value
                RaisePropertyChanged("DetailSource")
            End Set
        End Property
        Public Property CurrentRow() As DataRowView
            Get
                Return _currentRow
            End Get
            Set(ByVal value As DataRowView)
                If _currentRow Is value Then
                    Return
                End If
                _currentRow = value
                RaisePropertyChanged("CurrentRow")
                UpdateDetailSource()
            End Set
        End Property

        Public Sub New()
            DataSet = CreateData()
            MasterSource = DataSet.Tables("Company")
        End Sub

        Private Function CreateData() As DataSet
            Dim mdt As New DataTable("Company")
            mdt.Columns.Add(New DataColumn("Name", GetType(String)))
            mdt.Columns.Add(New DataColumn("ID", GetType(Integer)))
            mdt.Rows.Add("Ford", 4)
            mdt.Rows.Add("Nissan", 5)
            mdt.Rows.Add("Mazda", 6)

            Dim ddt As New DataTable("Models")
            ddt.Columns.Add(New DataColumn("Name", GetType(String)))
            ddt.Columns.Add(New DataColumn("MaxSpeed", GetType(Integer)))
            ddt.Columns.Add(New DataColumn("CompanyName", GetType(String)))
            ddt.Rows.Add("FordFocus", 400, "Ford")
            ddt.Rows.Add("FordST", 400, "Ford")
            ddt.Rows.Add("Note", 1000, "Nissan")
            ddt.Rows.Add("Mazda3", 1000, "Mazda")

            Dim ds As New DataSet("CM")
            ds.Tables.Add(mdt)
            ds.Tables.Add(ddt)
            Dim dr As New DataRelation("CompanyModels", mdt.Columns("Name"), ddt.Columns("CompanyName"))
            ds.Relations.Add(dr)

            Return ds
        End Function
        Private Sub UpdateDetailSource()
            Dim dv As New DataView(DataSet.Tables("Models"))
            dv.RowFilter = String.Format("CompanyName = '{0}'", CurrentRow("Name"))
            DetailSource = DataSet.Tables("Models")
        End Sub

        Public Event PropertyChanged As PropertyChangedEventHandler Implements INotifyPropertyChanged.PropertyChanged
        Private Sub RaisePropertyChanged(ByVal propName As String)
            RaiseEvent PropertyChanged(Me, New PropertyChangedEventArgs(propName))
        End Sub
    End Class
End Namespace